go/pserver/optimizer.go

```diff
@@ -1,42 +1,68 @@
 package pserver
 
-/*
-#include "optimizer.h"
-*/
+// #cgo pkg-config: protobuf
+// #cgo CFLAGS: -I ../../
+// //FIXME: ldflags contain "build" path
+// #cgo LDFLAGS: ../../build/go/pserver/cclient/libpaddle_go_optimizer.a -lstdc++
+// #include "paddle/optimizer/optimizer.h"
+// #include <stdlib.h>
+// #include <string.h>
 import "C"
+
 import (
 	"fmt"
 	"unsafe"
 )
 
-type optimizerType int
-
-const (
-	sgd optimizerType = iota
-)
-
 var nullPtr = unsafe.Pointer(uintptr(0))
 
 type optimizer struct {
 	opt *C.struct_paddle_optimizer
+	// used in GetParam, reconstruct Parameter from optimizer
+	ElementType ElementType
+}
+
+func cArrayToSlice(p unsafe.Pointer, len int) []byte {
+	if p == nullPtr {
+		return nil
+	}
+
+	// create a Go clice backed by a C array, reference:
+	// https://github.com/golang/go/wiki/cgo#turning-c-arrays-into-go-slices
+	//
+	// Go garbage collector will not interact with this data, need
+	// to be freed properly.
+	return (*[1 << 30]byte)(p)[:len:len]
 }
 
-func newOptimizer(t optimizerType, learning_rate float64) *optimizer {
+func newOptimizer(paramWithConfigs ParameterWithConfig) *optimizer {
 	o := &optimizer{}
-	o.opt = C.paddle_create_SGD_optimizer(C.double(learning_rate))
+	o.ElementType = paramWithConfigs.Param.ElementType
+	p := paramWithConfigs.Param
+	c := paramWithConfigs.Config
+	var cbuffer unsafe.Pointer
+	cbuffer_len := int(unsafe.Sizeof(p.Content[0])) * len(p.Content)
+	cbuffer = C.malloc(C.size_t(cbuffer_len))
+	C.memcpy(cbuffer, unsafe.Pointer(&p.Content[0]), C.size_t(cbuffer_len))
+	o.opt = C.paddle_create_optimizer((*C.uchar)(&c[0]), C.int(len(c)),
+		C.paddle_element_type(p.ElementType), cbuffer, C.int(len(p.Content)),
+		(*C.char)(nullPtr), 0)
```

Comment:
`c := paramWithConfigs.Config` is created but never used. Does this code compiles?